File: problem/pidigits/solution/node/6.js
/* The Computer Language Benchmarks Game
   https://salsa.debian.org/benchmarksgame-team/benchmarksgame/

   transliterated from Alexander Fyodorovʼs program by Isaac Gouy
*/

function pad(i, last) {
  var res = i.toString(), count;
  count = 10 - res.length;
  while (count > 0) {
    last ? res += ʼ ʼ : res = ʼ0ʼ + res;
    count--;
  }
  return res;
}

function calculatePi(arg) {
  let i = 0, ns = 0;

  let k = 0n;
  let k1 = 1n;
  let a = 0n;
  let d = 1n;
  let m = 0n;
  let n = 1n;
  let t = 0n;
  let u = 1n;

  while (true) {
    k += 1n;
    k1 += 2n;
    t = n << 1n;
    n *= k;
    a += t;
    a *= k1;
    d *= k1;

    if (a > n) {
      m = n * 3n + a;
      t = m / d;
      u = m % d + n;

      if (d > u) {
        ns = ns * 10 + Number(t);
        i += 1;

        let last = i >= arg;
        if (i % 10 == 0 || last) {

          console.log(pad(ns, last) + ʼ\t:ʼ + i);
          ns = 0;
        }

        if (last) break;

        a = (a - d * t) * 10n;
        n = n * 10n;
      }
    }
  }
}


calculatePi( +process.argv[2] || 30 );

File: problem/pidigits/solution/dartjit/6.dart
/* The Computer Language Benchmarks Game
   https://salsa.debian.org/benchmarksgame-team/benchmarksgame/

   contributed by Isaac Gouy, based on Alexander Fyodorov's Dart #2
*/

String pad(i, last) {
  var res = i.toString(), count;
  count = 10 - res.length;
  while (count > 0) {
    last ? res += ' ' : res = '0' + res;
    count--;
  }
  return res;
}

void calculatePi(arg) {
  var i = 0, ns = 0;

  final bigint_three = new BigInt.from(3);
  final bigint_ten = new BigInt.from(10);

  var k = BigInt.zero;
  var k1 = BigInt.one;
  var a = BigInt.zero;
  var d = BigInt.one;
  var m = BigInt.zero;
  var n = BigInt.one;
  var t = BigInt.zero;
  var u = BigInt.one;

  while (true) {
    k += BigInt.one;
    k1 += BigInt.two;
    t = n << 1;
    n *= k;
    a += t;
    a *= k1;
    d *= k1;

    if (a.compareTo(n) >= 0) {
      m = n * bigint_three + a;
      t = m ~/ d;
      u = m % d + n;

      if (d.compareTo(u) > 0) {
        ns = ns * 10 + t.toInt();
        i += 1;

        var last = i >= arg;
        if (i % 10 == 0 || last) {
          print(pad(ns, last) + '\t:$i');
          ns = 0;
        }

        if (last) break;

        a = (a - d * t) * bigint_ten;
        n = n * bigint_ten;
      }
    }
  }
}

void main(List<String> arguments) {
  calculatePi( int.parse(arguments[0]) );
}

